Diff2Flow: Aligning Diffusion Models with Flow-Matching

Koratahiu opened the pull request:
This PR integrates the **Diff2Flow** methodology, based on the paper "[Diff2Flow: Training Flow Matching Models via Diffusion Model Alignment](https://arxiv.org/abs/2506.02221)". 
This feature adapts a pre-trained diffusion model to a flow matching objective, unlocking significant performance improvements.

A new boolean option, `diff2flow`, has been added to the training configuration for both SD and SDXL models.

### Key Benefits & Features

-   **More Efficient Fine-Tuning:** The model converges much faster during training, especially with parameter-efficient methods like LoRA. This is because the model's pre-trained knowledge is directly leveraged without needing to re-learn a new prediction objective from scratch.
-   **Improved Sample Quality:** Diff2Flow naturally resolves the zero-terminal SNR issue often seen in eps-prediction models, leading to images with true blacks and whites instead of a washed-out or gray bias.
-   **Broad Compatibility:** The implementation is compatible with both `eps-pred` and `v-pred` models, seamlessly converting their outputs to the required velocity field.

### Implementation Details

-   The core logic is encapsulated in `modules/util/convert/diff2flow_util.py`.
-   The `enable_diff2flow()` function is called during model setup to dynamically add the necessary methods and buffers.
-   Samplers (`StableDiffusionSampler` and `StableDiffusionXLSampler`) now include a non-invasive dispatch to a dedicated Diff2Flow sampling loop that uses a simple Euler ODE solver.

### References 
The paper: https://arxiv.org/abs/2506.02221
The code: https://github.com/CompVis/diff2flow

### Checklist
- [x] Test on SD1.5
- [x] Test on SDXL
- [x] Test on v-pred model
- [ ] Maybe support inpainting? It might need some work.
- [ ] Remove or keep the loss weight function of diffusion (e.g., Min SNR). The referenced paper did not specify its usage.
- [x] Minimal Changes: The changes are minimal in all files except BaseStableDiffusionSetup and BaseStableDiffusionXLSetup, which might need some simplification.
- ~~Timestep sampling problem: The paper’s method samples a continuous time t uniformly from U[0,1] , then finds the corresponding diffusion timestep. However, we want to apply user input for timesteps. More tests are required.~~

Koratahiu committed:
Add Diff2Flow to SDXL and SD

Koratahiu committed:
Add Missing Logic and fix the zero div warning

Koratahiu committed:
fix zero div v2

Koratahiu committed:
Fix Typo in SDXL Sampling

Koratahiu committed:
Apply Continuous Timestep

Koratahiu committed:
Add Safeguard when Timestep == 0,1

Koratahiu committed:
Revert the clamping and add a better one

Koratahiu committed:
Change the logic of timestep sampling

Koratahiu committed:
Revert to original paper continuous time t

Koratahiu committed:
Revert to original paper continuous time t v2

Koratahiu committed:
Fix the miss placed

Koratahiu committed:
Sample continuous time t

Koratahiu committed:
Refine the timestep sampling and use _get_timestep_discrete for diff2flow logic

Goldenkoron commented:
I've been testing this both on eps and vpred models with SDXL and it's showing promise at improving those models in their native predictions even without using flow matching support in inference.

For example the eps model was able to do solid black images which it normally would never be able to do, and vpred ztsnr model got better with composition and darkness. I encourage others to test as well.

Koratahiu committed:
Merge branch 'Nerogar:master' into Diff2Flow

Koratahiu committed:
Rework diff2flow core logic and fix the bugs

Koratahiu committed:
Workaround for SNR-based weighting

Koratahiu committed:
Modify the SNR timesteps
This should be more robust since the old method resulted in NaNs for me, using Min SNR

Koratahiu committed:
Align OT timesteps with Diff2Flow logic
Thanks to @dxqb for the insight!

Koratahiu committed:
Pass betas/timestep for generalized offset noise

Koratahiu committed:
Rework: Minimal changes to (Base setup) files, reverse the target to (Noise - Data) in sampling/training to align with flow-matching/timesteps in OT.

O-J1 commented:
>  Maybe support inpainting? It might need some work.
> Remove or keep the loss weight function of diffusion (e.g., Min SNR). The referenced paper did not specify its usage.

Do you still think these need to occur?

Koratahiu commented:
I think inpainting is already supported in the latest commit, since we made it use the standard SDXL/SD path and only applied the diff2flow logic at the end (but still untested)

Regarding loss weights, they shouldn't be used with diff2flow (different loss, different logic)

Koratahiu commented:
Is this still needed? It’s basically ready, but who uses SDXL anymore?

miasik commented:
> Is this still needed? It’s basically ready, but who uses SDXL anymore?

I do
